File: Blind75/Find Element in Rotated Array/Find_Element_in_Rotated_Array.js
function findInRotatedArray(numbers, target){
    let l = 0
    let r = numbers.length - 1

    while(l <= r){
        let m = l + Math.floor((r-l)/2)

        if(numbers[m] == target) return m
        if (numbers[l] <= numbers[m]){
            // left half is sorted 
            if(target < numbers[m] && target >= numbers[l]){
                r = m - 1
            } else {
                l = m + 1
            }
        }
        else {
            // left half is sorted 
            if(target > numbers[m] && target <= numbers[r]){
                l = m + 1
            } else{
                r = m - 1
            }
        }
    }

    return -1
}

const testCases = [
    {numbers: [0,1,2,3,4], target: 2},
    {numbers: [2,3,4,0,1], target: 0},
    {numbers: [4], target: 2},
]
for (const {numbers, target} of testCases){
    console.log(findInRotatedArray(numbers, target))
}

File: Blind75/Find Element in Rotated Array/Find_Element_in_Rotated_Array.py
def findInRotatedArray(numbers: list[int], target: int):
    l = 0
    r = len(numbers) - 1
    
    while l <= r:
        m = l + (r - l) // 2

        if numbers[m] == target: return m
        if numbers[l] <= numbers[m]:
            # left side is sorted
            if target >= numbers[l] and target < numbers[m]:
                r = m - 1
            else: 
                l = m + 1
        else: 
            # right side is sorted 
            if target > numbers[m] and target <= numbers[r]:
                l = m + 1
            else:
                r = m - 1
        
    return -1

test_cases = [
    {"numbers": [0,1,2,3,4], "target": 2},
    {"numbers": [2,3,4,0,1], "target": 0},
    {"numbers": [4], "target": 2},
]

for case in test_cases:
    numbers = case["numbers"]
    target = case["target"]

    print(findInRotatedArray(numbers, target))